File: native-servers/android-device-server/src/androidTest/java/com/devicestream/server/util/JsonRpc.kt
package com.devicestream.server.util

import org.json.JSONObject

/**
 * JSON-RPC 2.0 response helpers.
 * Mirrors the iOS JsonRpc utility.
 */
object JsonRpc {

    fun successResponse(id: Any?, result: Any): String {
        val response = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("result", result)
            put("id", id)
        }
        return response.toString()
    }

    fun errorResponse(id: Any?, code: Int, message: String): String {
        val error = JSONObject().apply {
            put("code", code)
            put("message", message)
        }
        val response = JSONObject().apply {
            put("jsonrpc", "2.0")
            put("error", error)
            put("id", id)
        }
        return response.toString()
    }
}

File: native-servers/ios-xctest-server/XCTestServer/Util/JsonRpc.swift
import Foundation

/// JSON-RPC 2.0 types and serialization helpers.
enum JsonRpc {

    static func successResponse(id: Any?, result: Any) -> String {
        var response: [String: Any] = [
            "jsonrpc": "2.0",
            "result": result
        ]
        if let id = id {
            response["id"] = id
        }
        return serialize(response)
    }

    static func errorResponse(id: Any?, code: Int, message: String) -> String {
        var response: [String: Any] = [
            "jsonrpc": "2.0",
            "error": [
                "code": code,
                "message": message
            ] as [String: Any]
        ]
        if let id = id {
            response["id"] = id
        }
        return serialize(response)
    }

    private static func serialize(_ dict: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: dict, options: []),
              let str = String(data: data, encoding: .utf8) else {
            return #"{"jsonrpc":"2.0","error":{"code":-32603,"message":"Serialization error"}}"#
        }
        return str
    }
}
